Add await_connection to Integration Test API

ekswang opened the pull request:
Added wait_for_dataflow()

| | |
|:---|:---|
|**_Related Issue(s)_**|  [2642](https://github.com/nasa/fprime/issues/2642) |
|**_Has Unit Tests (y/n)_**|  |
|**_Documentation Included (y/n)_**|  |

---
## Change Description

There is a need to add a function in the fprime_test_api that allows tests to block or wait on data flow.

## Rationale

This function allows user to launch fit test calling fprime_test_api.wait_for_dataflow() and be able to bring up the flight software within a specified time duration.

## Testing/Review Recommendations

- Launch fprime-gds without the binary being loaded via one terminal
- Load the binary via a separate terminal and verify GSE is up and running
- Execute fit test
- While fit test is executing, kill the process that's running the binary
- Verify wait_for_dataflow() asserts when detecting no data flow for specified time duration
- Load the binary via a separate terminal and verify GSE is up and running
- Verify wait_for_dataflow() continues after detecting data flow

## Future Work

N/A

ekswang committed:
Update api.py

Added wait_for_dataflow()

ekswang committed:
Update api.py

- updated variable names

thomas-bc commented on src/fprime_gds/common/testing_fw/api.py:
```diff
@@ -234,6 +234,31 @@ def get_deployment(self):
+
+        history = self.get_telemetry_subhistory()
+        result = self.await_telemetry_count(
+            count, channels=channels, history=history, start=start, timeout=timeout
```
Does leaving `channels=None` accomplish the same result?

ekswang replied:
Yes, you are right. I just looked at ind_history_count() and it will default to search all if None is passed into the function. I will remove the unnecessary definition of channels.

thomas-bc commented on src/fprime_gds/common/testing_fw/api.py:
```diff
@@ -234,6 +234,31 @@ def get_deployment(self):
+        ch_dict = self.pipeline.dictionaries.channel_name
+        channels = list(ch_dict.keys())
+
+        history = self.get_telemetry_subhistory()
```
Why create a subhistory and not use the `get_telemetry_test_history` ?

ekswang replied:
Creating a subhistory will make sure that there's no residual/past (last instance/packet) data being returned due to the usage of getting the latest time, which gets the time from the last telemetry data,. Hence, await_telemetry_count function would have returned that object/data even when there's no data flow.

thomas-bc replied:
Sounds good - thanks for the explanation!

ekswang committed:
Update api.py

- Removed the need to define a list of channels to check (existing helper/api function already does it)

thomas-bc commented on src/fprime_gds/common/testing_fw/api.py:
```diff
@@ -234,6 +234,29 @@ def get_deployment(self):
+
+        history = self.get_telemetry_subhistory()
+        result = self.await_telemetry_count(
+            count, channels=channels, history=history, start=start, timeout=timeout
```
I think you missed changing to `channels=None` on your latest commit here

ekswang replied:
I set it to channels=None as default in the function. Just to be flexible for user.

thomas-bc replied:
Ohhh sorry my bad - thanks!
